### ISSUE TITLE
New Feature Request: filter() Interpolation for applying a function to a list of elements

### ISSUE THREAD

**Issue by dkheyman** [hashicorp/terraform#17468]
I am running into an issue of how to apply a function to a list. No such helper function really exists. I come from a Python background where applying functions to a list is quite rudimentary using filter(), and map(), and lambdas.

The use case for this would be, for example, to apply a filter to values in a map, according to a value, for example. There does not seem to be an easy way to extract data from a map on a condition yet. Referenced issues were brought up with no quality workaround yet.

One could also come up with any sort of conditional in the filter function. 
### Terraform Version

```
Terraform v0.11.3
```

### Terraform Configuration Files

```
[in main.tf environment]
my_map = {
  'elem_1' = {
    'type' = 'ec2',
    'name' = 'ec2-iam-role'
  },
  'elem_2' = {
    'type' = 'ec2',
    'name' = 'ec2-other-iam-role'
  }
}

```
This returns a list of the values with key 'type' matching 'ec2'. Very simple in Python, but I don't believe something like this exists currently.

Python: 
```
matching_elements = (filter(lambda x: x["type"] == 'ec2', my_map.values())). 
```

### Sample Invocation

```
locals {
  x = ""
}

matching_elements = "${filter(lookup(local.x, "type", "") == 'ec2', values(var.my_map))}"
```

### Expected Behavior

The expected behavior would be that matching_elements would leave you with a list of elements that have that conditional function applied to them. I would love something like this, as well as a map() function that takes a function and applies it to a list.

### References

https://github.com/hashicorp/terraform/issues/12453

**Comment by dkheyman** [comment]
This is in addition to the ```matchkeys``` function, because matchkeys only allows equality, or match. This would expand it to really do any kind of lookup or conditional.

**Comment by apparentlymart** [comment]
Hi @dkheyman,

A future major release will include our improved configuration language interpreter which has a construct that I think addresses your use-case here:

```hcl
  # Not yet implemented, and details may change before release
  matching_elements = [for v in var.my_map: v if v.type == "ec2"]
```

As someone who is familiar with Python, I think you'll recognize this as being a list comprehension. The Terraform syntax for it is "reversed" compared to Python to allow for future implementation of code completion features (need to know what type `v` is before you type `v.type`) but the functionality here is equivalent to a Python expression like this:

```python
 matching_elements = [x for x in my_map.values() if x["type"] == "ec2"]
```

More details about this new feature can be seen in [its specification](https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/spec.md#for-expressions). Integrating this new implementation (which has a number of other features in addition to this) is the current focus for the Terraform team at HashiCorp.

**Comment by dkheyman** [comment]
That sounds very useful. Looking forward to addressing the pains of list comprehension in a new release. Thanks!

**Comment by bgshacklett** [comment]
@apparentlymart I know that you've all been working very hard on HCL, but there are a number of open PRs for interpolation functions that would make end-users lives significantly easier and allow for modules with significantly more functionality. Can you give us any idea of how long it's going to be before we see a version of Terraform with HCL2 released?

**Comment by tdmalone** [comment]
>  I would love something like this, as well as a map() function that takes a function and applies it to a list.

This was mentioned at the bottom of the OP for this issue. I would love to see this. Should a separate issue be created for it?

**Comment by apparentlymart** [comment]
Both `map` and `filter` can be implemented using the _`for` expression_ feature I described earlier. There's a more detailed writeup about this now in [one of our articles on forthcoming features in v0.12](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each), including an example of a `map`-like use-case.

**Comment by apparentlymart** [comment]
Hi all,

As discussed over in #8439, the `for` expression feature I mentioned earlier is now merged in master and will be included in the forthcoming v0.12.0 final release. Therefore I'm going to close this out. Thanks for this feature request, and thanks for your patience while we laid the groundwork to implement it.

**Comment by whindes** [comment]
One work around to filter maps without using "for loops".  If ther is a better way, please chime in.
`zipmap(var.filter_keylist
,matchkeys(values(var.some_map.value),keys(var.some_map.value)
,var.filter_keylist))`
